test/e2e_node/device_plugin.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e_node
+
+import (
+	"time"
+
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+	dp "k8s.io/kubernetes/pkg/kubelet/cm/deviceplugin"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// Serial because the test restarts Kubelet
+var _ = framework.KubeDescribe("Device Plugin [Feature:DevicePlugin] [Serial] [Disruptive]", func() {
+	f := framework.NewDefaultFramework("device-plugin-errors")
+
+	Context("DevicePlugin", func() {
+		By("Enabling support for Device Plugin")
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.FeatureGates[string(features.DevicePlugins)] = true
+		})
+
+		BeforeEach(func() {
+
+		})
+
+		AfterEach(func() {
+
+		})
+
+		It("Checks that Kubelet handles device plugin registration correctly.", func() {
+
+			By("Start stub device plugin")
+			devs := []*pluginapi.Device{
+				{ID: "Dev1", Health: pluginapi.Healthy},
+				{ID: "Dev2", Health: pluginapi.Healthy},
+			}
+
+			p1 := dp.NewDevicePluginStub(devs, pluginapi.DevicePluginPath+"dp.1")
+			err := p1.Start()
+			framework.ExpectNoError(err)
+
+			By("Wait for node is ready")
+			Eventually(func() int {
+				nodeList := framework.GetReadySchedulableNodesOrDie(f.ClientSet)
+				return len(nodeList.Items)
+			}, 30*time.Second, framework.Poll).Should(Equal(1))
+
+			By("Register resources")
+			resourceName := "mock1.com/resource"
+			p1.Register(pluginapi.KubeletSocket, resourceName)
+
```

Comment:
The general model looks good! @ScorpioCPH do you think if we can extend the test to test the same set of functionalities as the current gpu device plugin e2e_node does? This includes:
- Tests that after device plugin registration, a user pod requesting the registered resource can run on the node.
- Tests that Kubelet correctly configures user pod container runtime environment according to the AllocResponse from the stub device plugin. For this part, you may need to extend the stub device plugin to take a allocFunc during its creation.
- Tests that kubelet restart doesn't disrupt running pods.
- Confirming that after a kubelet and pod restart, GPU assignment is kept.
- And following steps in test/e2e_node/gpu_device_plugin.go ...

Reply:
Sure, I will extend these steps, maybe this week :)